exporter/internal/queue/batcher.go

```diff
@@ -0,0 +1,259 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package queue // import "go.opentelemetry.io/collector/exporter/internal/queue"
+
+import (
+	"context"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/exporter/exporterbatcher"
+	"go.opentelemetry.io/collector/exporter/internal"
+)
+
+type batch struct {
+	ctx     context.Context
+	req     internal.Request
+	idxList []uint64
+}
+
+type Batcher struct {
+	cfg            exporterbatcher.Config
+	mergeFunc      exporterbatcher.BatchMergeFunc[internal.Request]
+	mergeSplitFunc exporterbatcher.BatchMergeSplitFunc[internal.Request]
+
+	queue      Queue[internal.Request]
+	maxWorkers int
+	numWorkers atomic.Int64
+	exportFunc func(context.Context, internal.Request) error
+
+	batchListMu      sync.Mutex
+	batchList        []batch
+	lastFlushed      time.Time
+	timer            *time.Timer
+	shutdownCh       chan bool
+	allocateWorkerCh chan bool
+
+	stopWG sync.WaitGroup
+}
+
+func NewBatcher(cfg exporterbatcher.Config, queue Queue[internal.Request],
+	maxWorkers int, exportFunc func(context.Context, internal.Request) error,
+	mergeFunc exporterbatcher.BatchMergeFunc[internal.Request],
+	mergeSplitFunc exporterbatcher.BatchMergeSplitFunc[internal.Request]) *Batcher {
```

Comment:
@dmitryax do you remember why the merge and split are not part of the Request?

Reply:
Do you mean having them as methods? I don't remember precisely. Probably because it involves many `Request` objects.